### PR TITLE
Update resnet18.py

### PR DESCRIPTION
**Pull request by aysha2016**
Command-Line Arguments: Modify the script to accept command-line arguments, making it more versatile. Users can specify the input image file, working precision, and model type as command-line arguments.

### CODE REVIEW

File: examples/2_ResNet18/resnet18.py [hunk 1]
```diff
@@ -94,7 +105,7 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
 
     print_top_results(output)
 
-
+"""
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:108]: This is creating errors

File: examples/2_ResNet18/resnet18.py [hunk 2]
```diff
@@ -132,3 +143,24 @@ def print_top_results(output: torch.Tensor) -> None:
 
     rn_model = initialize(torch_precision)
     run_model(rn_model, np_precision)
+"""
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:146]: This isn't indented properly.

File: examples/2_ResNet18/resnet18.py [hunk 3]
```diff
@@ -118,10 +129,10 @@ def print_top_results(output: torch.Tensor) -> None:
...
         )
-
+"""
 
 if __name__ == "__main__":
-    np_precision = np.float32
+ """   np_precision = np.float32
 
     if np_precision == np.float32:
         torch_precision = torch.float32
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:138]: Dead code shouldn't be left in - it has been superseded by the below.

File: examples/2_ResNet18/resnet18.py [hunk 4]
```diff
@@ -63,6 +65,15 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
             ),
         ]
     )
+      # Show the input image
+    plt.figure(figsize=(6, 6))
+    plt.imshow(input_image)
+    plt.title("Input Image")
+    plt.axis('off')
+    plt.show()
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:73]: This is great as long as the user has a system that can execute `plt.show()`. Perhaps there could be an option for show with it being switched off by default. We can assume if a user has downloaded and provided an image they know what it looks like.

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Hi @aysha2016 Thank you for using FTorch and opening up a community contribution.

Unfortunately I was unable to run this code due to a few suprious `"""` and indentation errors. I have noted a few, but not all, in your code.

Please could you update to something that runs on your system and then let us know, and we will then conduct a more in-depth review.

File: examples/2_ResNet18/resnet18.py [hunk 5]
```diff
@@ -35,23 +35,8 @@ def initialize(precision: torch.dtype) -> torch.nn.Module:
...
-    Parameters
-    ----------
-    model: torch.nn.Module
-        Pretrained model to run.
-    precision: type
-        NumPy data type to save input tensor.
-    """
-    # Transform image into the form expected by the pre-trained model, using the mean
-    # and standard deviation from the ImageNet dataset
-    # See: https://pytorch.org/vision/0.8/models.html
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:53]: The docstrings are important to help users understand the code, so please leave them in (this applies to the below docstrings as well.)

File: examples/2_ResNet18/resnet18.py [hunk 6]
```diff
@@ -73,11 +58,11 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
     )  # type: np.typing.NDArray
 
     # Save data as binary
-    np_input.tofile("data/image_tensor.dat")
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:76]: Please keep the images and data files in the `data/` subdirectory.

File: examples/2_ResNet18/resnet18.py [hunk 7]
```diff
@@ -92,43 +77,36 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
...
-    categories = np.genfromtxt(cats_filename, dtype=str, delimiter="\n")
-
-    # Show top categories per image
-    top5_prob, top5_catid = torch.topk(probabilities, 5)
-    print("\nTop 5 results:\n")
-    for i in range(top5_prob.size(0)):
-        cat_id = top5_catid[i]
-        print(
-            f"{categories[cat_id]} (id={cat_id}): probability = {top5_prob[i].item()}"
-        )
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:120]: This function is useful for allowing users to view and evaluate the results, so please keep it in the code.

File: examples/2_ResNet18/resnet18.py [hunk 8]
```diff
@@ -92,43 +77,36 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
...
+    elif args.precision == "fp64":
+        np_precision = np.float64
         torch_precision = torch.float64
     else:
-        raise ValueError("`np_precision` must be of type `np.float32` or `np.float64`")
+        raise ValueError("Precision must be 'fp32' or 'fp64'")
 
     rn_model = initialize(torch_precision)
-    run_model(rn_model, np_precision)
+    run_model(rn_model, np_precision, args.image_path)
```
Comment (jatkinson1000) [inline examples/2_ResNet18/resnet18.py:112]: Please include some checks/help for the image_path variable (see https://docs.python.org/3/library/argparse.html#example for a good example)

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Thank you for pushing your changes to this PR @aysha2016 and thank you for addressing some of the comments from the previous review.

Unfortunately you have removed some things that we would prefer to be left in the code such as documentation and evaluation functions. If you could put these back and update the PR it would be appreciated.

We have updated the main branch of the code since you opened this PR, so you may want to consider syncing your fork to get the latest code.

Finally, please also update the local README for this exercise to specify how the user could/should run the code with your argparse additions.

Many thanks.

**Review by jatkinson1000 — COMMENTED** [review]
Thank you, this is starting to look better.
Please let us know in the comment and by requesting a re-review when you are ready.
Please add some instructions to the README in exercise 2 to describe how to run the code in its new format.

Brief comments:
- please keep data files in the `data/` directory
- Consider updating your fork to bring it in line with the latest changes

**Comment by jatkinson1000** [comment]
Thanks @aysha2016 
I can't see your updates here, but it looks like you have made them in the `main` branch of [your fork](https://github.com/aysha2016/FTorch).
If you merge `main` into the `aysha2016-patch-1` branch they should appear here.

File: examples/2_ResNet18/README.md [hunk 9]
```diff
@@ -35,13 +36,15 @@ source venv/bin/activate
...
-You can check that everything is working by running `resnet18.py`:
+You can run the provided resnet18.py script from the command line:
 
 ```
-python3 resnet18.py
+python resnet18.py /path/to/your/image.jpg --precision fp32 --model_type resnet18
 ```
 
-When using single precision, it should produce the result:
+Replace /path/to/your/image.jpg with the path to the image you want to classify.
```
Comment (jatkinson1000) [inline examples/2_ResNet18/README.md:45]: Could you make the example run on `data/dog.jpg` to ensure that it matches the example results below, and then move your explanation of how to specify other images to the "Further Options" section at the end of this file please?

**Review by jatkinson1000 — CHANGES_REQUESTED** [review]
Thank you @aysha2016
The precision command line argument is looking great!
A few general comments below, and specific points noted in the code.

This PR is currently [failing the automated checks](https://github.com/Cambridge-ICCS/FTorch/actions/runs/7114466328/job/19368931377?pr=63) as it does not conform to the [black](https://black.readthedocs.io/en/stable/index.html) code style.
You can remedy this by running
```
pip install black
black myfile.py
```
to reformat python file `myfile.py` and then committing the result.

Please see in the code for other comments.

File: examples/2_ResNet18/resnet18.py [hunk 10]
```diff
@@ -120,15 +105,35 @@ def print_top_results(output: torch.Tensor) -> None:
...
+        default="fp32",
+        help="Working precision (default: fp32)",
+    )
+    parser.add_argument(
+        "--model_type",
+        choices=["resnet18"],
+        default="resnet18",
+        help="Model type (default: resnet18)",
+    )
+    args = parser.parse_args()
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/resnet18.py]: Many thanks.  I have removed the --model_type command line argument.

File: examples/2_ResNet18/resnet18.py [hunk 11]
```diff
@@ -120,15 +105,35 @@ def print_top_results(output: torch.Tensor) -> None:
...
+    elif args.precision == "fp64":
+        np_precision = np.float64
         torch_precision = torch.float64
     else:
-        raise ValueError("`np_precision` must be of type `np.float32` or `np.float64`")
+        raise ValueError("Precision must be 'fp32' or 'fp64'")
 
     rn_model = initialize(torch_precision)
-    run_model(rn_model, np_precision)
+    run_model(rn_model, np_precision, args.image_path,"data/categories.txt")
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/resnet18.py]: I have made the suggested change

File: examples/2_ResNet18/resnet18.py [hunk 12]
```diff
@@ -36,22 +36,7 @@ def initialize(precision: torch.dtype) -> torch.nn.Module:
...
-    model: torch.nn.Module
-        Pretrained model to run.
-    precision: type
-        NumPy data type to save input tensor.
-    """
-    # Transform image into the form expected by the pre-trained model, using the mean
-    # and standard deviation from the ImageNet dataset
-    # See: https://pytorch.org/vision/0.8/models.html
-    image_filename = "data/dog.jpg"
+def run_model(model: torch.nn.Module, precision: type, image_filename: str, categories_filename: str) -> None:
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/resnet18.py]: I have made the suggested change

File: examples/2_ResNet18/resnet18.py [hunk 13]
```diff
@@ -36,22 +36,7 @@ def initialize(precision: torch.dtype) -> torch.nn.Module:
...
-    Parameters
-    ----------
-    model: torch.nn.Module
-        Pretrained model to run.
-    precision: type
-        NumPy data type to save input tensor.
-    """
-    # Transform image into the form expected by the pre-trained model, using the mean
-    # and standard deviation from the ImageNet dataset
-    # See: https://pytorch.org/vision/0.8/models.html
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/resnet18.py]: included again .

File: examples/2_ResNet18/resnet18.py [hunk 14]
```diff
@@ -73,11 +58,11 @@ def run_model(model: torch.nn.Module, precision: type) -> None:
     )  # type: np.typing.NDArray
 
     # Save data as binary
-    np_input.tofile("data/image_tensor.dat")
+    np_input.tofile("image_tensor.dat")
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/resnet18.py]: I have made the suggested change

File: examples/2_ResNet18/README.md [hunk 15]
```diff
@@ -35,13 +36,15 @@ source venv/bin/activate
 pip install -r requirements.txt
 ```
 
-You can check that everything is working by running `resnet18.py`:
+You can run the provided resnet18.py script from the command line:
 
 ```
-python3 resnet18.py
+python resnet18.py /path/to/your/image.jpg --precision fp32 --model_type resnet18
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/README.md]: I have made the suggested change.

File: examples/2_ResNet18/README.md [hunk 16]
```diff
@@ -53,8 +56,7 @@ Pomeranian (id=259): probability = 0.005621383432298899
 Great Pyrenees (id=257): probability = 0.004652013536542654
 ```
 
-To save the pretrained ResNet-18 model to TorchScript run the modified version of the
-`pt2ts.py` tool :
+To save the pretrained ResNet-18 model to TorchScript run the modified version of the pt2ts.py` tool :
```
[the remark this answers is not in the document]
Comment (aysha2016) [reply examples/2_ResNet18/README.md:59]: Many thanks for addressing the issue. I have made changes as per your suggestion.

Aysha

**Comment by jatkinson1000** [comment]
Closing this as stale.

**Pull request closed without merging**